docs/notebooks/async_bo_demo.py

```diff
@@ -0,0 +1,326 @@
+initial_observations = objective(initial_query_points.numpy(), sleep=False)
+initial_data = Dataset(query_points=initial_query_points, observations=tf.constant(initial_observations, dtype=tf.float64))
+
+initial_data
+
+# %% [markdown]
+# Model definition
+
+# %%
+import gpflow
+from trieste.models import create_model
+
+from trieste.models.gpflow.config import GPflowModelConfig
+
+def build_model(data):
+    variance = tf.math.reduce_variance(data.observations)
+    kernel = gpflow.kernels.RBF(variance=variance)
+    gpr = gpflow.models.GPR(data.astuple(), kernel, noise_variance=1e-5)
+    gpflow.set_trainable(gpr.likelihood, False)
+
+    return GPflowModelConfig(**{
+        "model": gpr,
+        "optimizer": gpflow.optimizers.Scipy(),
+        "optimizer_args": {
+            "minimize_args": {"options": dict(maxiter=100)},
+        },
+    })
+
+
+# %% [markdown]
+# Define a worker that runs a single observation in a separate process. Worker reads next point from the points queue, makes an observation, and inserts observed data into the observations queue.
+
+# %%
+import psutil
+
+def observer_proc(points_queue, observations_queue, cpu_id):
+    pid = os.getpid()
+    
+    current_process = psutil.Process()
+    current_process.cpu_affinity([cpu_id])
+    print(f"Process {pid}: set CPU to {cpu_id}", flush=True)
+    
+    while True:
+        point_to_observe = points_queue.get()
+        if (point_to_observe is None):
+            return
+        
+        print(f"Process {pid}: Observer : observing data at point {point_to_observe}", flush=True)
+        new_observation = objective(point_to_observe, sleep=enable_sleep_delays)
+        new_data = (point_to_observe, new_observation)
+        
+        print(f"Process {pid}: Observer : observed data {new_data}", flush=True)
+
+        observations_queue.put(new_data)
+
+
+# %% [markdown]
+# Set how many workers can run simultaneously, and iterations for sync case
+
+# %%
+num_workers = 3
+num_iterations = 10
+
+# %% [markdown]
+# ## Normal sync BO with LP
+#
+# We use LP acquisition function, and run all batch points in parallel. For that we setup the worker processes, as many as there are 
+
+# %%
+from trieste.acquisition import LocalPenalizationAcquisitionFunction
+from trieste.acquisition.rule import AsyncEGO, EfficientGlobalOptimization
+from trieste.ask_tell_optimization import AskTellOptimizer
+from multiprocessing import Manager, Process
+
+# setup Ask Tell BO
+model_spec = build_model(initial_data)
+model = create_model(model_spec)
+
+local_penalization_acq = LocalPenalizationAcquisitionFunction(search_space, num_samples=2000)
+local_penalization_rule = EfficientGlobalOptimization(
+    num_query_points=num_workers, builder=local_penalization_acq
+)
+
+sync_bo = AskTellOptimizer(search_space, initial_data, model, local_penalization_rule)
+
+
+# thread-safe queues
+m = Manager()
+points_q = m.Queue()
+observations_q = m.Queue()
+observer_processes = []
+pid = os.getpid()
+
```

Comment:
I dont really understand any of the the multi processing stuff. This will need to be better documented in the final version!